[HUDI-3919] [UBER] Support out of order rollback blocks in AbstractHoodieLogRecordReader

suryaprasanna opened the pull request:
## What is the purpose of the pull request

*This pull request adds support for out of order rollback blocks in AbstractHoodieLogRecordReader.*

## Brief change log

  - *Change the iteration logic in AbstractHoodieLogRecordReader to support handling various multiwriter scenarios.*
  - *Fix issue when rollback blocks are farther from their target blocks*
  - *Include test case to verify out of order rollback blocks*

## Verify this pull request

*This pull request change is already covered by existing tests, and also modified a new test case to verify the changes*.

## Committer checklist

 - [ ] Has a corresponding JIRA in PR title & commit
 
 - [ ] Commit message is descriptive of the change
 
 - [ ] CI is green

 - [ ] Necessary doc changes done or have another open PR
       
 - [ ] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA.

nsivabalan commented:
@alexeykudinkin : can you review this

prashantwason commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -218,7 +221,45 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
           readerSchema, readBlocksLazily, reverseReader, bufferSize, enableRecordLookups, keyField, internalSchema);
 
+      /**
+       * Traversal of log blocks from log files can be done in two directions.
```
Please simplify this comment and provide only the current logic.

prashantwason commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -245,97 +286,53 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
             continue;
           }
         }
+        if (logBlock.getBlockType().equals(CORRUPT_BLOCK)) {
```
Can we handle this too in the switch that follows? Having a common way to handle the various block types is easier to understand as per code flow.

prashantwason commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -245,97 +286,53 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
+          continue;
+        }
+
+        // Rollback blocks contain information of instants that are failed, collect them in a set..
```
This comments seems more relevant to where the rollback block is being handled later.

prashantwason commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -245,97 +286,53 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
       }
+
+      int numBlocksRolledBack = 0;
+      // This is a reverse traversal on the collected data blocks.
```
collected data and delete blocks.

How is this reverse traversal? Isnt the for-loop a forward traversal?

prashantwason commented on hudi-common/src/test/java/org/apache/hudi/common/functional/TestHoodieLogFormat.java:
```diff
@@ -839,20 +839,24 @@ public void testAvroLogRecordReaderWithRollbackTombstone(ExternalSpillableMap.Di
     writer.appendBlock(dataBlock);
 
     // Write 2
+    header = new HashMap<>();
```
header.clear() also works instead of allocating a new hashmap each time.

prashantwason commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -218,7 +221,45 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
           logFilePaths.stream().map(logFile -> new HoodieLogFile(new Path(logFile))).collect(Collectors.toList()),
           readerSchema, readBlocksLazily, reverseReader, bufferSize, enableRecordLookups, keyField, internalSchema);
```
Lets also remove the readBlocksLazily argument as it now required to be always true.

prashantwason commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -218,7 +221,45 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
           logFilePaths.stream().map(logFile -> new HoodieLogFile(new Path(logFile))).collect(Collectors.toList()),
           readerSchema, readBlocksLazily, reverseReader, bufferSize, enableRecordLookups, keyField, internalSchema);
 
```
Lets also remove the reverseReader as it is no longer supported.

suryaprasanna commented:
@prashantwason Addressed the review comments. Removed readBlocksLazily and reverseReader flags from AbstractHoodieLogRecordReader and log file reader classes.
Although, I have not removed following configs COMPACTION_LAZY_BLOCK_READ_ENABLE, COMPACTION_REVERSE_LOG_READ_ENABLE from HoodieWriteConfig object.

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -232,7 +251,7 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
         )) {
           // hit a block with instant time greater than should be processed, stop processing further
-          break;
+          continue;
```
why continue. blocks greater than latest known instant time can be skipped altogether right?

nsivabalan commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/HoodieLogFileReader.java:
```diff
@@ -93,35 +89,26 @@ public HoodieLogFileReader(FileSystem fs, HoodieLogFile logFile, Schema readerSc
     this.enableRecordLookups = enableRecordLookups;
     this.keyField = keyField;
     this.internalSchema = internalSchema == null ? InternalSchema.getEmptyInternalSchema() : internalSchema;
-    if (this.reverseReader) {
```
why we are removing the reverse reader ? can you help me understand

suryaprasanna replied:
My understanding is that when iterating in reverse order there is an issue when we encounter corrupt block. We cannot jump across the corrupt block since we dont have the block size stored at the end for them. So, we end up ignoring all the blocks older than the corrupt block. 
That is a reason for removing the reverseReader lookup, since it cannot be handled.
It becomes more complicated when introducing log compaction. There we need to move the compacted blocks to a different slot. So, it is not straight forward traversal. So, removing this logic to reduce the complexity involved.

suryaprasanna replied:
Please let me know, what do you think?

nsivabalan commented on hudi-common/src/test/java/org/apache/hudi/common/functional/TestHoodieLogFormat.java:
```diff
@@ -414,7 +411,7 @@ public void testHugeLogFileWrite() throws IOException, URISyntaxException, Inter
     byte[] dataBlockContentBytes = getDataBlock(DEFAULT_DATA_BLOCK_TYPE, records, header).getContentBytes();
     HoodieLogBlock.HoodieLogBlockContentLocation logBlockContentLoc = new HoodieLogBlock.HoodieLogBlockContentLocation(new Configuration(), null, 0, dataBlockContentBytes.length, 0);
-    HoodieDataBlock reusableDataBlock = new HoodieAvroDataBlock(null, Option.ofNullable(dataBlockContentBytes), false,
+    HoodieDataBlock reusableDataBlock = new HoodieAvroDataBlock(null, Option.ofNullable(dataBlockContentBytes),
```
do we have tests for test out the multi-writer scenario. i.e rollback log block is appended after few other valid log blocks? if not, can we add one.

suryaprasanna replied:
Sure, I will add them.

nsivabalan reviewed (commented):
hey Surya, thanks for the patch. Wondering, for single writer scenario, do we think we can retain old behavior. only for multi-writer and minor log compactions, we might have to take the new route.

suryaprasanna commented on hudi-common/src/main/java/org/apache/hudi/common/table/log/AbstractHoodieLogRecordReader.java:
```diff
@@ -218,7 +221,45 @@ protected synchronized void scanInternal(Option<KeySpec> keySpecOpt) {
+       *  This becomes more complicated if we have compacted blocks, which are data blocks created using log compaction.
+       *  TODO: Include support for log compacted blocks. https://issues.apache.org/jira/browse/HUDI-3580
+       *
+       *  To solve this do traversal twice.
```
Two traversals is needed to support the multiwriter scenarios where we can have rollback way away from the original block it is targeting. With minor compaction it becomes more tricky since we can have compacted blocks comprising of other compacted blocks. So, tackling the multiwriter scenarios with this PR first.

suryaprasanna commented:
> hey Surya, thanks for the patch. Wondering, for single writer scenario, do we think we can retain old behavior. only for multi-writer and minor log compactions, we might have to take the new route.

Does that mean we need to pass multiwriter enabled flag to AbstractHoodieLogRecordReader and using that flag toggle the logic between one traversal and two traversals?

nsivabalan commented on hudi-client/hudi-client-common/src/main/java/org/apache/hudi/table/action/rollback/BaseRollbackHelper.java:
```diff
@@ -210,7 +210,7 @@ protected Map<HoodieLogBlock.HeaderMetadataType, String> generateHeader(String c
     header.put(HoodieLogBlock.HeaderMetadataType.TARGET_INSTANT_TIME, commit);
     header.put(HoodieLogBlock.HeaderMetadataType.COMMAND_BLOCK_TYPE,
-        String.valueOf(HoodieCommandBlock.HoodieCommandBlockTypeEnum.ROLLBACK_PREVIOUS_BLOCK.ordinal()));
+        String.valueOf(HoodieCommandBlock.HoodieCommandBlockTypeEnum.ROLLBACK_BLOCK.ordinal()));
```
thanks!

nsivabalan commented:
Closing in favor of https://github.com/apache/hudi/pull/5958
